topk-js/src/expr.rs

```diff
@@ -0,0 +1,343 @@
+    match &self.expr {
+      Expression::LogicalExpression { expr } => Self {
+        expr: Expression::LogicalExpression {
+          expr: expr.divide(&LogicalExpression::Literal {
+            value: Value::F64(value),
+          }),
+        },
+      },
+      _ => unreachable!("divide can only be called on logical expressions"),
+    }
+  }
+
+  #[napi]
+  pub fn and(&self, other: Expr) -> Expr {
+    match &self.expr {
+      Expression::TextExpression { expr } => {
+        if let Expression::TextExpression { expr: other_expr } = &other.expr {
+          Self {
+            expr: Expression::TextExpression {
+              expr: expr.and(other_expr),
+            },
+          }
+        } else {
+          unreachable!("and can only be called on text expressions")
+        }
+      }
+      Expression::LogicalExpression { expr } => Self {
+        expr: Expression::LogicalExpression {
+          expr: expr.and(&other.expr.into()),
+        },
+      },
+      _ => unreachable!("and can only be called on logical expressions"),
+    }
+  }
+
+  #[napi]
+  pub fn or(&self, other: Expr) -> Expr {
+    match &self.expr {
+      Expression::TextExpression { expr } => {
+        if let Expression::TextExpression { expr: other_expr } = &other.expr {
+          Self {
+            expr: Expression::TextExpression {
+              expr: expr.or(other_expr),
+            },
+          }
+        } else {
+          unreachable!("or can only be called on text expressions")
+        }
+      }
+      Expression::LogicalExpression { expr } => Self {
+        expr: Expression::LogicalExpression {
+          expr: expr.or(&other.expr.into()),
+        },
+      },
+      _ => unreachable!("or can only be called on logical expressions"),
+    }
+  }
+
+  #[napi]
+  pub fn starts_with(&self, value: String) -> Expr {
+    match &self.expr {
+      Expression::LogicalExpression { expr } => Self {
+        expr: Expression::LogicalExpression {
+          expr: expr.starts_with(&LogicalExpression::Literal {
+            value: Value::String(value),
+          }),
+        },
+      },
+      _ => unreachable!("starts_with can only be called on logical expressions"),
+    }
+  }
+}
+
+impl FromNapiValue for Expr {
+  unsafe fn from_napi_value(
+    env: napi::sys::napi_env,
+    value: napi::sys::napi_value,
+  ) -> Result<Self, napi::Status> {
+    let object = Object::from_napi_value(env, value)?;
+    let expr: Expression = object.get("expr")?.ok_or_else(|| {
+      napi::Error::new(
+        napi::Status::GenericFailure,
+        "Field object missing 'expr' property".to_string(),
+      )
+    })?;
+
+    Ok(Self { expr })
+  }
+}
+
+#[napi]
+pub fn field(name: String) -> Expr {
+  Expr::create_field(name)
+}
+
+#[napi(object)]
+#[derive(Debug, Clone)]
+pub struct MatchOptions {
+  pub field: String,
+  pub weight: f64,
+}
+
+#[napi(js_name = "match")]
+pub fn match_(token: String, options: Option<MatchOptions>) -> Expr {
```

Comment:
tip: you can either do `match_` or use rust's fancy syntax `r#match`

Reply:
looks like napi is complaining:

<img width="860" alt="Screenshot 2025-04-02 at 13 49 53" src="https://github.com/user-attachments/assets/d73870c8-6a66-4b1b-bd75-b2ebbc4d1bef" />

Reply:
freaking procedural macros!